services/datamanager/builtin/builtin_sync_test.go

```diff
@@ -695,10 +710,39 @@ func TestArbitraryFileUpload(t *testing.T) {
 			case <-closeCtx.Done():
 			}
 
+			// Call upload to dataset.
+			datasetBytes := []byte("hi")
+			datasetsMetadata := []string{"dataset1"}
+			tagsMetadata := []string{"tag1"}
+			mimeType := v1.MimeType_MIME_TYPE_IMAGE_PNG
+			if tc.uploadToDataset {
+				t.Log("calling upload to a dataset")
+				// Add the bytes to the file contents.
+				fileContents = append(fileContents, datasetBytes...)
+				timeoutCtx, timeoutFn := context.WithTimeout(context.Background(), time.Second*5)
+				defer timeoutFn()
+				for {
+					if err = b.UploadBinaryDataToDataset(context.Background(), datasetBytes, datasetsMetadata, tagsMetadata, mimeType, nil); err == nil {
+						break
+					}
+
+					if timeoutCtx.Err() != nil {
+						t.Log("timed out waiting for mocked cloud connection")
+						t.FailNow()
+					}
+					time.Sleep(time.Millisecond * 50)
+				}
+			}
+
 			waitUntilNoFiles(additionalPathsDir)
+			waitUntilNoFiles(filepath.Join(captureDir, datasync.DatasetDir))
 			if !tc.serviceFail {
 				// Validate first metadata message.
-				test.That(t, uploadCount.Load(), test.ShouldEqual, 1)
+				expectedUploadedCount := 1
```

Comment:
Can we make this a variable that runs per test. Meaning, TestA has 3 expectedUploadedCount, TestB has 2...